### PR TITLE
Add BackupBucket and BackupEntry controllers

### LINKED ISSUE #219: Extend Gardener Extension Provider for `onmetal` with `BackupBucket` and `BackupEntry` Controllers
## Feature 

**Is your feature request related to a problem? Please describe.**
Currently, the existing Gardener extension provider for `onmetal` doesn't support the `BackupBucket` and `BackupEntry` controllers. This limits the provider's functionality and might cause unnecessary complications when dealing with backup management on `onmetal` deployments.

**Describe the solution you'd like**
We propose the addition of two new controllers, `BackupBucket` and `BackupEntry`, to the `onmetal` Gardener extension provider. This feature would allow the system to create, update, and delete backup buckets and entries, providing more comprehensive and straightforward backup management capabilities.

**Additional context**
As Gardener is increasingly used in combination with `onmetal` environments, it's necessary to continue enhancing its capabilities and align them with the existing Gardener features for other environments.

The two new controllers could provide the following functionalities:

1. **BackupBucket Controller**: Manages the lifecycle of backup buckets.

2. **BackupEntry Controller**: Manages the lifecycle of backup entries.

We believe that by implementing these controllers, it would provide a more holistic approach to manage backups and restore operations of `Shoots` running on `onmetal` `Seed` clusters.

**Tasks Involved**
1. Design the architecture of the two new controllers in line with Gardener's existing architectural style.
2. Implement the new controllers.
3. Test the implementation in various scenarios to ensure stability and reliability.
4. Document the new features for end users and other developers.

**Open Questions**
1. Are there specific Gardener guidelines that we need to follow while implementing these controllers?
2. Are there limitations on the OnMetal environment that could affect this implementation?

### PR DESCRIPTION
**Pull request by ushabelgur**
# Proposed Changes

- Add BackupBucket Controller: Manages the lifecycle of backup buckets.
- Add BackupEntry Controller: Manages the lifecycle of backup entries.
- Add tests for both the controllers.

Fixes #219

### CODE REVIEW

**Review by afritzler — CHANGES_REQUESTED** [review]

File: pkg/controller/backupbucket/actuator.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+	log.V(2).Info("Reconciling backupBucket")
+
+	onmetalClient, namespace, err := onmetal.GetOnmetalClientAndNamespaceFromSecretRef(ctx, a.Client(), &backupBucket.Spec.SecretRef)
+	if err != nil {
+		return fmt.Errorf("failed to get onmetal client and namespace from cloudprovider secret: %w", err)
+	}
+
+	// If the generated secret in the backupbucket status not exists that means
+	// no backupbucket exists and it need to be created.
+	if backupBucket.Status.GeneratedSecretRef == nil {
```
[the remark this answers is not in the document]
Comment (ushabelgur) [reply pkg/controller/backupbucket/actuator.go:54]: How will I get shoot Namespace ?

[the remark this answers is not in the document]
Comment (afritzler) [reply pkg/controller/backupbucket/actuator.go:54]: The `BackupBucket` resource is in the `Shoot` `Namespace`

File: example/30-backupentry.yaml [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+# namespace: base64(namespace)
+# kubeconfig: base64(kubecconfig)
+---
+apiVersion: extensions.gardener.cloud/v1alpha1
+kind: BackupEntry
+metadata:
+  name: shoot--foobar--onmetal--sd34f
+spec:
+  type: onmetal
+  region: eu-west-1
```
[the remark this answers is not in the document]
Comment (ushabelgur) [reply example/30-backupentry.yaml]: Done

File: example/30-backupbucket.yaml [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+# namespace: base64(namespace)
+# kubeconfig: base64(kubecconfig)
+---
+apiVersion: extensions.gardener.cloud/v1alpha1
+kind: BackupBucket
+metadata:
+  name: cloud--onmetal--fg2d6
+spec:
+  type: onmetal
+  region: northeurope
```
[the remark this answers is not in the document]
Comment (ushabelgur) [reply example/30-backupbucket.yaml]: Done

File: pkg/controller/backupbucket/actuator.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ushabelgur) [reply pkg/controller/backupbucket/actuator.go:56]: Storing secret in the backupBucket.Spec.SecretRef.Namespace with Bucket access secret and endpoint deatils

File: pkg/controller/backupbucket/backupbucket.go [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+		Spec: storagev1alpha1.BucketSpec{
+			BucketClassRef: &corev1.LocalObjectReference{
+				Name: a.backupBucketConfig.BucketClassName,
+			},
+		},
+	}
+
+	if _, err := controllerutil.CreateOrPatch(ctx, onmetalClient, bucket, nil); err != nil {
+		return fmt.Errorf("failed to create or patch backup bucket %s: %w", client.ObjectKeyFromObject(bucket), err)
+	}
```
[the remark this answers is not in the document]
Comment (ushabelgur) [reply pkg/controller/backupbucket/backupbucket.go:51]: Incorporated suggested changes

File: pkg/onmetal/types.go [hunk 5]
```diff
@@ -43,6 +43,10 @@ const (
 	// MachineControllerManagerProviderOnmetalImageName is the name of the MachineController onmetal image.
 	MachineControllerManagerProviderOnmetalImageName = "machine-controller-manager-provider-onmetal"
 
+	// AccessKeyID is a constant for the key in a cloud provider secret and backup secret that holds the Bucket access key id.
+	AccessKeyID = "accessKeyID"
+	// SecretAccessKey is a constant for the key in a cloud provider secret and backup secret that holds the Bucket secret access key.
+	SecretAccessKey = "secretAccessKey"
```
[the remark this answers is not in the document]
Comment (ushabelgur) [reply pkg/onmetal/types.go:53]: added endpoint key

File: pkg/controller/backupbucket/actuator.go [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+		if err := validateConfiguration(a.backupBucketConfig); err != nil {
+			return fmt.Errorf("error validating configuration: %w", err)
+		}
+		err := a.ensureBackupBucket(ctx, namespace, onmetalClient, backupBucket)
```
[the remark this answers is not in the document]
Comment (ushabelgur) [reply pkg/controller/backupbucket/actuator.go]: done

File: pkg/controller/backupbucket/actuator.go [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+		if err != nil {
+			return fmt.Errorf("failed to create backup bucket: %w", err)
+		}
+
+		accessSecret := &corev1.Secret{}
+		if err := onmetalClient.Get(ctx, client.ObjectKey{Namespace: namespace, Name: bucketAccess.SecretRef.Name}, accessSecret); err != nil {
+			return fmt.Errorf("error getting bucket access secret")
+		}
+
+		log.V(2).Info("Successfully reconciled backupBucket")
```
Comment (afritzler) [inline pkg/controller/backupbucket/actuator.go:71]: Please move this to the end and log out "Reconciled BackupBucket".

File: pkg/controller/backupbucket/actuator.go [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+
+func newActuator(backupBucketConfig *controllerconfig.BackupBucketConfig) backupbucket.Actuator {
+	return &actuator{
+		backupBucketConfig: backupBucketConfig,
+	}
+}
+
+// Reconcile implements backupbucket.Actuator
+func (a *actuator) Reconcile(ctx context.Context, log logr.Logger, backupBucket *extensionsv1alpha1.BackupBucket) error {
+	log.V(2).Info("Reconciling backupBucket")
```
Comment (afritzler) [inline pkg/controller/backupbucket/actuator.go:47]:
```suggestion
	log.V(2).Info("Reconciling BackupBucket")
```

File: pkg/controller/backupbucket/actuator.go [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+
+	onmetalClient, namespace, err := onmetal.GetOnmetalClientAndNamespaceFromSecretRef(ctx, a.Client(), &backupBucket.Spec.SecretRef)
+	if err != nil {
+		return fmt.Errorf("failed to get onmetal client and namespace from cloudprovider secret: %w", err)
+	}
+
+	bucket := &storagev1alpha1.Bucket{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      backupBucket.Name,
+			Namespace: namespace,
```
Comment (afritzler) [inline pkg/controller/backupbucket/actuator.go:92]: `BackupBucket` is a cluster scoped object and has no namespace.

File: pkg/controller/backupbucket/actuator.go [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+		// update backupBucket status with secret ref
+		return a.updateBackupBucketStatus(backupBucket, accessSecret.Data, bucketAccess.Endpoint, ctx)
+
+	}
+
+	return nil
+}
+
+func (a *actuator) Delete(ctx context.Context, log logr.Logger, backupBucket *extensionsv1alpha1.BackupBucket) error {
+	log.V(2).Info("Deleting Bucket")
```
Comment (afritzler) [inline pkg/controller/backupbucket/actuator.go:82]:
```suggestion
	log.V(2).Info("Deleting BackupBucket")
```

File: pkg/controller/backupbucket/actuator.go [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      backupBucket.Name,
+			Namespace: namespace,
+		},
+	}
+	if err = onmetalClient.Delete(ctx, bucket); err != nil {
+		return fmt.Errorf("failed to delete backup bucket : %v", err)
+	}
+
+	log.V(2).Info("Deleted Bucket")
```
Comment (afritzler) [inline pkg/controller/backupbucket/actuator.go:99]:
```suggestion
	log.V(2).Info("Deleted BackupBucket")
```

File: pkg/controller/backupbucket/actuator.go [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+	}
+
+	bucket := &storagev1alpha1.Bucket{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      backupBucket.Name,
+			Namespace: namespace,
+		},
+	}
+	if err = onmetalClient.Delete(ctx, bucket); err != nil {
+		return fmt.Errorf("failed to delete backup bucket : %v", err)
```
Comment (afritzler) [inline pkg/controller/backupbucket/actuator.go:96]:
```suggestion
		return fmt.Errorf("failed to delete backup bucket: %v", err)
```

File: pkg/controller/backupentry/actuator.go [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+	a.client = client
+	return nil
+}
+
+func (a *actuator) GetETCDSecretData(_ context.Context, _ logr.Logger, _ *extensionsv1alpha1.BackupEntry, backupSecretData map[string][]byte) (map[string][]byte, error) {
+	return backupSecretData, nil
+}
+
+func (a *actuator) Delete(ctx context.Context, log logr.Logger, backupEntry *extensionsv1alpha1.BackupEntry) error {
+
```
Comment (afritzler) [inline pkg/controller/backupentry/actuator.go:47]: Please remove newline.

File: pkg/controller/backupbucket/actuator.go [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+
+		accessSecret := &corev1.Secret{}
+		if err := onmetalClient.Get(ctx, client.ObjectKey{Namespace: namespace, Name: bucketAccess.SecretRef.Name}, accessSecret); err != nil {
+			return fmt.Errorf("error getting bucket access secret")
+		}
+
+		log.V(2).Info("Successfully reconciled backupBucket")
+
+		// update backupBucket status with secret ref
+		return a.updateBackupBucketStatus(backupBucket, accessSecret.Data, bucketAccess.Endpoint, ctx)
```
Comment (afritzler) [inline pkg/controller/backupbucket/actuator.go:74]: Move this method above the log output and pass only the `backupBucket` and the `accessSecret` to it. `ctx` should be the first argument. The `bucketAccess.Endpoint` can be get from the `backupBucket` object.

File: pkg/controller/backupbucket/actuator.go [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+	}
+
+	// If the generated secret in the backupbucket status not exists that means
+	// no backupbucket exists and it need to be created.
+	if backupBucket.Status.GeneratedSecretRef == nil {
+		if err := validateConfiguration(a.backupBucketConfig); err != nil {
+			return fmt.Errorf("error validating configuration: %w", err)
+		}
+
+		bucketAccess, err := a.ensureBackupBucket(ctx, namespace, onmetalClient, backupBucket)
```
Comment (afritzler) [inline pkg/controller/backupbucket/actuator.go:61]: Use a single line `if err := ...` and remove the return parameter. The pointer to the `backupBucket` will have already all the information we need.

File: pkg/controller/backupbucket/actuator_test.go [hunk 16]
```diff
@@ -0,0 +1,170 @@
...
+		Expect(k8sClient.Create(ctx, bucketAccesSecret)).To(Succeed())
+
+		By("patching backup bucket with available state and access details")
+		bucketBase := bucket.DeepCopy()
+		bucket.Status.State = storagev1alpha1.BucketStateAvailable
+		bucket.Status.Access = &storagev1alpha1.BucketAccess{
+			SecretRef: &corev1.LocalObjectReference{
+				Name: "my-bucket-secret",
+			},
+			Endpoint: "endpoint-efef-ihfbd-ssadd.s3.storage",
```
Comment (afritzler) [inline pkg/controller/backupbucket/actuator_test.go:102]: Please don't use s3 in the endpoint.

File: pkg/controller/backupbucket/actuator_test.go [hunk 17]
```diff
@@ -0,0 +1,170 @@
...
+		}
+		Expect(k8sClient.Status().Patch(ctx, bucket, client.MergeFrom(bucketBase))).To(Succeed())
+
+		By("ensuring that bucket is created and Available")
+		Eventually(Object(bucket)).Should(SatisfyAll(
+			HaveField("Status.State", storagev1alpha1.BucketStateAvailable),
+			HaveField("Status.Access.SecretRef.Name", "my-bucket-secret"),
+			HaveField("Status.Access.Endpoint", "endpoint-efef-ihfbd-ssadd.s3.storage"),
```
Comment (afritzler) [inline pkg/controller/backupbucket/actuator_test.go:110]: Please don't use s3 in the endpoint.

**Review by afritzler — CHANGES_REQUESTED** [review]

File: pkg/controller/backupbucket/actuator.go (hunk 14 above)
Comment (Ashughorla) [reply pkg/controller/backupbucket/actuator.go:74]: Here `backupBucket` object is of `backupBucket *extensionsv1alpha1.BackupBucket`. It doesn't have `Endpoint`.
`ensureBackupBucket()` is returning access details from `onmetal` bucket.

Comment (afritzler) [reply pkg/controller/backupbucket/actuator.go:74]: Sorry for the confusion, I of course meant the `Bucket` which you create above.

File: pkg/controller/backupbucket/actuator.go (hunk 7 above)
Comment (Ashughorla) [reply pkg/controller/backupbucket/actuator.go]: Addressed.

File: pkg/controller/backupbucket/actuator.go (hunk 8 above)
Comment (Ashughorla) [reply pkg/controller/backupbucket/actuator.go]: Addressed.

File: pkg/controller/backupbucket/actuator.go (hunk 10 above)
Comment (Ashughorla) [reply pkg/controller/backupbucket/actuator.go]: Addressed.

File: pkg/controller/backupbucket/actuator.go (hunk 11 above)
Comment (Ashughorla) [reply pkg/controller/backupbucket/actuator.go]: Addressed.

File: pkg/controller/backupbucket/actuator.go (hunk 12 above)
Comment (Ashughorla) [reply pkg/controller/backupbucket/actuator.go]: Addressed.

File: pkg/controller/backupentry/actuator.go (hunk 13 above)
Comment (Ashughorla) [reply pkg/controller/backupentry/actuator.go]: Addressed.

File: pkg/controller/backupbucket/actuator.go (hunk 14 above)
Comment (Ashughorla) [reply pkg/controller/backupbucket/actuator.go]: Addressed.

File: pkg/controller/backupbucket/actuator.go (hunk 15 above)
Comment (Ashughorla) [reply pkg/controller/backupbucket/actuator.go]: Addressed.

File: pkg/controller/backupbucket/actuator_test.go (hunk 16 above)
Comment (Ashughorla) [reply pkg/controller/backupbucket/actuator_test.go]: Addressed.

File: pkg/controller/backupbucket/actuator_test.go (hunk 17 above)
Comment (Ashughorla) [reply pkg/controller/backupbucket/actuator_test.go]: Addressed.

File: pkg/controller/backupbucket/backupbucket.go [hunk 18]
```diff
@@ -0,0 +1,135 @@
...
+			return true, nil
+		}
+		return false, nil
+	}); err != nil {
+		return fmt.Errorf("could not determine status of backup bucket")
+	}
+
+	accessSecret := &corev1.Secret{}
+	if err := onmetalClient.Get(ctx, client.ObjectKey{Namespace: namespace, Name: bucket.Status.Access.SecretRef.Name}, accessSecret); err != nil {
+		return fmt.Errorf("error bucket access secret")
```
Comment (afritzler) [inline pkg/controller/backupbucket/backupbucket.go:65]: Can you please wrap and return the error itself here.

File: pkg/controller/backupbucket/backupbucket.go [hunk 19]
```diff
@@ -0,0 +1,135 @@
...
+	}
+	return nil
+}
+
+func isBucketAccessDetailsAvailable(bucket *storagev1alpha1.Bucket) bool {
+	return bucket.Status.Access != nil && bucket.Status.Access.SecretRef != nil && bucket.Status.Access.Endpoint != ""
+}
+
+// updateBackupBucketStatus updates backupBucket status with access secretRef
+func (a *actuator) updateBackupBucketStatus(ctx context.Context, backupBucket *extensionsv1alpha1.BackupBucket, secretData map[string][]byte, endpoint string) error {
```
Comment (afritzler) [inline pkg/controller/backupbucket/backupbucket.go:79]: I would call this method `patchBackupBucketStatus`.

File: pkg/controller/backupbucket/backupbucket.go [hunk 20]
```diff
@@ -0,0 +1,135 @@
...
+
+	secretAccessKey, ok := secretData[onmetal.BucketSecretAccessKey]
+	if !ok {
+		return fmt.Errorf("missing %q field in secret", onmetal.BucketSecretAccessKey)
+	}
+
+	accessSecretData := map[string][]byte{}
+	accessSecretData[onmetal.AccessKeyID] = []byte(accessKeyID)
+	accessSecretData[onmetal.SecretAccessKey] = []byte(secretAccessKey)
+	accessSecretData[onmetal.Endpoint] = []byte(endpoint)
```
Comment (afritzler) [inline pkg/controller/backupbucket/backupbucket.go:97]: We might want to use here constants which are defined by the [`gardener/etcd-backup-restore`](https://github.com/gardener/etcd-backup-restore) project. That way we ensure, that we store the credentials in the correct keys.

File: pkg/onmetal/types.go [hunk 21]
```diff
@@ -43,6 +43,16 @@ const (
 	// MachineControllerManagerProviderOnmetalImageName is the name of the MachineController onmetal image.
 	MachineControllerManagerProviderOnmetalImageName = "machine-controller-manager-provider-onmetal"
 
+	// AccessKeyID is a constant for the key in a onmetal bucket access secret that holds the Bucket access key id.
+	BucketAccessKeyID = "AWS_ACCESS_KEY_ID"
+	// AwsSecretAccessKey is a constant for the key in a onmetal bucket access secret that holds the Bucket secret access key.
+	BucketSecretAccessKey = "AWS_SECRET_ACCESS_KEY"
```
Comment (afritzler) [inline pkg/onmetal/types.go:49]: @lukasfrank do we have somewhere constants for those access secret keys? Or are those coming from `ceph` directly?

**Review by afritzler — CHANGES_REQUESTED** [review]

File: pkg/controller/backupbucket/backupbucket.go (hunk 20 above)
Comment (ushabelgur) [reply pkg/controller/backupbucket/backupbucket.go:97]: @afritzler I am not finding any constants in [gardener/etcd-backup-restore](https://github.com/gardener/etcd-backup-restore) repo. Can you point out from where I can refer?

File: pkg/controller/backupbucket/backupbucket.go (hunk 19 above)
Comment (ushabelgur) [reply pkg/controller/backupbucket/backupbucket.go]: method renamed

File: pkg/controller/backupbucket/backupbucket.go (hunk 20 above)
Comment (afritzler) [reply pkg/controller/backupbucket/backupbucket.go:97]: I just had a look and there don't seem to be any. :(

File: pkg/controller/backupbucket/actuator.go [hunk 22]
```diff
@@ -0,0 +1,86 @@
...
+	// no backupbucket exists and it need to be created.
+	if backupBucket.Status.GeneratedSecretRef == nil {
+		if err := validateConfiguration(a.backupBucketConfig); err != nil {
+			return fmt.Errorf("failed to validate configuration: %w", err)
+		}
+
+		if err := a.ensureBackupBucket(ctx, namespace, onmetalClient, backupBucket); err != nil {
+			return fmt.Errorf("failed to ensure backupbucket: %w", err)
+		}
+		log.V(2).Info("Reconciled BackupBucket")
```
Comment (afritzler) [inline pkg/controller/backupbucket/actuator.go:62]: I would move this log line out of the `if` statement.

File: pkg/controller/backupbucket/actuator_test.go [hunk 23]
```diff
@@ -0,0 +1,170 @@
...
+			Endpoint: "endpoint-efef-ihfbd-ssadd.storage",
+		}
+		Expect(k8sClient.Status().Patch(ctx, bucket, client.MergeFrom(bucketBase))).To(Succeed())
+
+		By("ensuring that bucket is created and Available")
+		Eventually(Object(bucket)).Should(SatisfyAll(
+			HaveField("Status.State", storagev1alpha1.BucketStateAvailable),
+			HaveField("Status.Access.SecretRef.Name", "my-bucket-secret"),
+			HaveField("Status.Access.Endpoint", "endpoint-efef-ihfbd-ssadd.storage"),
+		))
```
Comment (afritzler) [inline pkg/controller/backupbucket/actuator_test.go:111]: This test is not necessary as we trust the `onmetal-api` API server to handle this correctly.

File: pkg/controller/backupbucket/actuator_test.go [hunk 24]
```diff
@@ -0,0 +1,170 @@
...
+
+		Eventually(Get(bucket)).Should(Succeed())
```
Comment (afritzler) [inline pkg/controller/backupbucket/actuator_test.go:113]: We are checking the `bucket` on top already - it should exist at this point. This line can be removed.

File: pkg/controller/backupbucket/actuator_test.go [hunk 25]
```diff
@@ -0,0 +1,170 @@
...
+
+		By("ensuring backupbucket is delete successfully")
+		Expect(k8sClient.Delete(ctx, backupBucket)).To(Succeed())
+
+		By("waiting for the bucket to be gone")
+		Eventually(Get(bucket)).Should(Satisfy(apierrors.IsNotFound))
+	})
+
+	It("should check backup bucket configuration", func(ctx SpecContext) {
+
```
Comment (afritzler) [inline pkg/controller/backupbucket/actuator_test.go:148]: Please remove the new line here.

File: pkg/controller/backupbucket/actuator_test.go [hunk 26]
```diff
@@ -0,0 +1,170 @@
...
+		Expect(err).To(MatchError("backupBucketConfig must not be empty"))
+
+		By("validating bucketclassname is not empty")
+		err = validateConfiguration(config)
+		Expect(err).To(MatchError("BucketClassName is mandatory"))
+
+		config.BucketClassName = "foo"
+
+		By("validating backupbucketconfig is valid")
+		ret := validateConfiguration(config)
```
Comment (afritzler) [inline pkg/controller/backupbucket/actuator_test.go:165]: Please do a single line check here.

File: pkg/controller/backupbucket/actuator_test.go [hunk 27]
```diff
@@ -0,0 +1,170 @@
...
+
+	It("should check backup bucket configuration", func(ctx SpecContext) {
+
+		By("creating backupbucketconfig")
+		config := &controllerconfig.BackupBucketConfig{
+			BucketClassName: "",
+		}
+
+		By("validating backupbucket config")
+		err := validateConfiguration(nil)
```
Comment (afritzler) [inline pkg/controller/backupbucket/actuator_test.go:155]: Please do a single line check here.

File: pkg/controller/backupbucket/actuator_test.go [hunk 28]
```diff
@@ -0,0 +1,170 @@
...
+		Expect(err).To(MatchError("backupBucketConfig must not be empty"))
+
+		By("validating bucketclassname is not empty")
+		err = validateConfiguration(config)
```
Comment (afritzler) [inline pkg/controller/backupbucket/actuator_test.go:159]: Please do a single line check here.

File: pkg/controller/backupbucket/backupbucket.go [hunk 29]
```diff
@@ -0,0 +1,135 @@
...
+				Name: a.backupBucketConfig.BucketClassName,
+			},
+		},
+	}
+	//create onmetal bucket
+	if _, err := controllerutil.CreateOrPatch(ctx, onmetalClient, bucket, nil); err != nil {
+		return fmt.Errorf("failed to create or patch backup bucket %s: %w", client.ObjectKeyFromObject(bucket), err)
+	}
+	//wait for bucket creation
+	if err := wait.PollUntilWithContext(ctx, 1*time.Second, func(ctx context.Context) (done bool, err error) {
```
Comment (afritzler) [inline pkg/controller/backupbucket/backupbucket.go:53]: Can we use here a backoff polling like we are doing in the CCM: https://github.com/onmetal/cloud-provider-onmetal/blob/main/pkg/cloudprovider/onmetal/load_balancer.go#L187

File: pkg/controller/backupentry/actuator_test.go [hunk 30]
```diff
@@ -0,0 +1,151 @@
...
+		Expect(err).NotTo(HaveOccurred())
+
+		ctrl = gomock.NewController(GinkgoT())
+		mockS3ObjectLister = NewMocks3ObjectLister(ctrl)
+
+		objectLister = mockS3ObjectLister
+	})
+
+	It("should delete Backupentry", func(ctx SpecContext) {
+
```
Comment (afritzler) [inline pkg/controller/backupentry/actuator_test.go:62]: Please remove the newline.

File: pkg/controller/backupentry/actuator_test.go [hunk 31]
```diff
@@ -0,0 +1,151 @@
...
+				BucketPoolRef: &corev1.LocalObjectReference{
+					Name: "my-bucket-pool",
+				},
+				Tolerations: []commonv1alpha1.Toleration{
+					{
+						Key:      "key",
+						Operator: "Equal",
+						Value:    "value",
+					},
+				},
```
Comment (afritzler) [inline pkg/controller/backupentry/actuator_test.go:86]: We can remove the tolerations here.

File: pkg/controller/backupentry/actuator_test.go [hunk 32]
```diff
@@ -0,0 +1,151 @@
...
+			},
+		}
+		Expect(k8sClient.Create(ctx, bucket)).Should(Succeed())
+
+		Eventually(Object(bucket)).Should(SatisfyAll(
+			HaveField("Spec.BucketClassRef", &corev1.LocalObjectReference{
+				Name: "test-bucket-class",
+			}),
+		))
```
Comment (afritzler) [inline pkg/controller/backupentry/actuator_test.go:95]: No need to check here for the `classRef` as we are creating it above.

**Review by afritzler — CHANGES_REQUESTED** [review]

File: pkg/controller/backupbucket/actuator.go (hunk 22 above)
Comment (ushabelgur) [reply pkg/controller/backupbucket/actuator.go]: done

File: pkg/controller/backupbucket/actuator_test.go (hunk 24 above)
Comment (ushabelgur) [reply pkg/controller/backupbucket/actuator_test.go]: removed

File: pkg/controller/backupbucket/actuator_test.go (hunk 26 above)
Comment (ushabelgur) [reply pkg/controller/backupbucket/actuator_test.go]: done

File: pkg/controller/backupbucket/actuator_test.go (hunk 28 above)
Comment (ushabelgur) [reply pkg/controller/backupbucket/actuator_test.go]: done

File: pkg/controller/backupentry/actuator_test.go (hunk 30 above)
Comment (ushabelgur) [reply pkg/controller/backupentry/actuator_test.go:60]: done

File: pkg/controller/backupentry/actuator_test.go (hunk 31 above)
Comment (ushabelgur) [reply pkg/controller/backupentry/actuator_test.go]: done

File: pkg/controller/backupentry/actuator_test.go (hunk 32 above)
Comment (ushabelgur) [reply pkg/controller/backupentry/actuator_test.go]: removed

File: pkg/controller/backupbucket/backupbucket.go (hunk 29 above)
Comment (ushabelgur) [reply pkg/controller/backupbucket/backupbucket.go]: Modified to use wait.ExponentialBackoffWithContext

File: pkg/controller/backupbucket/actuator_test.go (hunk 23 above)
Comment (ushabelgur) [reply pkg/controller/backupbucket/actuator_test.go]: removed

File: pkg/controller/backupbucket/actuator_test.go (hunk 27 above)
Comment (ushabelgur) [reply pkg/controller/backupbucket/actuator_test.go]: done

**Review by afritzler — APPROVED** [review]

**Pull request merged**